src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp

```diff
@@ -128,11 +128,16 @@ void ShenandoahHeapRegion::make_young_maybe() {
    case _humongous_start:
    case _humongous_cont:
      if (affiliation() != YOUNG_GENERATION) {
-       if (is_old()) {
-         ShenandoahHeap::heap()->old_generation()->decrement_affiliated_region_count();
+       ShenandoahHeap* heap = ShenandoahHeap::heap();
+       if (heap->mode()->is_generational()) {
+         if (is_old()) {
+           heap->old_generation()->decrement_affiliated_region_count();
+         }
        }
        set_affiliation(YOUNG_GENERATION);
-       ShenandoahHeap::heap()->young_generation()->increment_affiliated_region_count();
+       if (heap->mode()->is_generational()) {
+         heap->young_generation()->increment_affiliated_region_count();
+       }
```

Comment:
I assume the strict order here is on account of assertion checks in some of these methods? Otherwise, one might simplify it by moving the affiliation setting out from the sandwich of the count adjustments which are conditional on generational mode.

Ideally the count for young gen in non-generational mode would work for the generational case, and we could just adjust the old count like now and drop the adjustment of young count (lines 138-140) entirely, because they are covered by line 137.

Reply:
We should be able to remove one of the mode checks. We don't maintain affiliated region counts in non-generational modes. There is a `TODO` for this: https://github.com/openjdk/shenandoah/blob/master/src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp#L872

I think this PR makes it easier to tackle that `TODO`, but it will be a separate PR.